README.md
QGIS Web Client 2 Demo Application
==================================

**Consult the [ChangeLog](https://github.com/qgis/qwc2-demo-app/blob/master/ChangeLog.md) for a summary of changes between releases**

Quick Start
-----------

Clone the repository with the --recursive option to automatically clone submodules.
* If you are a developer, use

        git clone --recursive https://github.com/qgis/qwc2-demo-app.git

* If you are an end-user and want to clone the latest stable release, use

        git clone --recursive --branch stable https://github.com/qgis/qwc2-demo-app.git

Install NodeJS, if needed, from [here](https://nodejs.org/en/blog/release/v0.12.7/).

We highly recommend using [yarn](https://yarnpkg.com/) to manage the dependencies instead of `npm`.

Start the development application locally:

    yarn install
    yarn start

The application runs at `http://localhost:8081` afterwards. You can change the address and port
by tweaking the `start` script in `package.json`, i.e.

    "start": "npm run themesconfig && webpack-dev-server --progress --colors --host <ip_addr> --port <port>"


Configuration
--------------

### Application

There are five main configuration files:

 * `js/appConfig.js`: This file configures which components and which map
 projections are compiled into the application bundle, the human readable labels
 for the EPSG codes as well as the available locales.
 Changing this file will require a re-deployment of the application.

 * `config.json`: This file controls the plugin configuration (whether they are
 enabled in desktop/mobile plus plugin specific configuration) and also stores
 some global properties. This file can be changed without re-deploying the application.

 * `styleConfig.js`: This file contains variables for the main styling properties
 (mainly colors). Changing this file will require a re-deployment of the application.

 * `themesConfig.json`: This file configures the themes which will be available
 in the application, see below for details.

 * `index.html`: This is the main entry point of the application. Among other things,
 the window title is set here. This file can be changed without re-deploying the application.

In addition, there are some additional important files (changing any of these
files will require a re-depolyment of the application):

 * `js/Help.jsx`: This component is displayed when selecting Help in the
 Application Menu. It can be customized as desired.

 * `js/SearchProviders.js`: This file contains the implementation of the search
 providers. The key names for the providers specified in the `module.exports`
 in this file can be listed under `searchProviders` in `themesConfig.json` to
 control which search servies are available for which themes.

 * `assets/css/qwc2.css`: This is a global style sheet which can contain any
 desired style tweaks.

### Map projection configuration

The map projection is configured individually for each theme in `themesConfig.json` (see below).
Remember to register the projection you wish to use in `js/appConfig.js`.
Note that the leaflet backend currently does not support projections other than `EPSG:3857`.

### Themes

Themes and map scales are configured in `themesConfig.json`.

Configuration format:
```
{
  "themes": {
    "items": [
      {
        "url": "<http://localhost/wms/theme>",
        "title": "<Custom theme title>",            // optional, use WMS title if not set
        "thumbnail": "<filename>",                  // optional image file in assets/img/mapthumbs, if not set uses WMS GetMap to generate the thumbnail and stores it in assets/img/mapthumbs
        "attribution": "<Attribution>",             // optional theme attribution
        "attributionUrl": "<attribution URL>",      // optional theme attribution URL
        "default": true,                            // optional, set this as the initial theme
        "scales": [25000, 10000, 5000, 2500],       // optional, custom map scales, defaults to defaultScales (see below)
        "printScales": [25000, 10000, 5000, 2500],  // optional, confined list of available print scales, defaults to defaultPrintScales (see below)
        "printResolutions": [150, 300, 600],        // optional, confined list of abailable print resolutions, defaults to defaultPrintResolutions (see below)
        "printGrid": [                              // optional, list of grid intervals to use for various scales when printing.
            {"s": 10000, x: 1000, y: 1000},         //   Keep this list sorted in descending order by scale (s)
            {"s": 1000, x: 100, y: 100},            //   In this example, {x: 100, y: 100} will be used for 1000 <= scale < 10000
            ...                                     //   If not specified, defaultPrintGrid will be usd (see below)
        ],
        "printLabelForSearchResult": "<labelid>"    // optional, a labelid in the print composition where to insert the label of the selected search result
        "printLabelConfig": {                       // optional, configuration of input textareas for print composition labels.
            "labelId": {
                "rows": 4,
                "maxLength: "40"
              },
              ...
        }
        "extent": [xmin, ymin, xmax, ymax],         // optional custom extent which overrides extent from WMS capabilities
        "tiled": true,                              // optional, use tiled WMS (default is false)
        "format": "image/png",                      // optional, the image format to use in the WMS request, defaults to image/png
        "backgroundLayers": [                       // optional background layers
          {
            "name": "<background layer name>",      // background layer name from list below
            "printLayer": "<qgis layer name>",      // optional, name of a qgis layer to use as equivalent background layer when printing
            "visibility": true                      // optional initial visibility on topic selection
          }
        ],
        "searchProviders": ["<search provider>"],   // optional search providers
        "mapCrs: "EPSG:3857",                       // optional, the map projection, defaults to EPSG:3857
        "additionalMouseCrs": ["<epsg code>"]       // optional list of additional CRS for mouse position (map projection and WGS84 are listed by default). Make sure proj defs are loaded in js/appConfig.js.
        "watermark": {                              // optional, configuration of watermark to place on raster-export images
          "text": "<watermark text>",
          "texpadding": "1",                        // optional, padding between text and frame, in points
          "fontsize": "14",                         // optional, font size
          "fontfamily": "sans",                     // optional, font family
          "fontcolor": "#0000FF",                   // optional, font color
          "backgroundcolor": "#FFFFFF",             // optional, background color of the frame
          "framecolor": "#000000",                  // optional, color of the frame border
          "framewidth": 1                           // optional, width of the frame border, in pixels
        },
        "collapseLayerGroupsBelowLevel": <level>    // optional, layer tree level below which to initially collapse groups. If unspecified, groups are not initially collapsed.
        "skipEmptyFeatureAttributes": true          // optional, whether to skip empty feature attributes in the identify results. Default is false.
      }
    ],
    "groups": [                                     // optional, nested groups
      {
        "title": "<Group title>",
        "items": [
          {
            "url": "<http://localhost/wms/group_theme>"
          }
        ]
        "groups": [
          // subgroups
        ]
      }
    ],
    "backgroundLayers": [                           // optional list of background layers for themes
      {
        "name": "<background layer name>",          // referenced by themes
        "title": "<Background layer title>",
        "thumbnail": "<filename>",                  // optional image file in assets/img/mapthumbs, use default.jpg if not set
        ...                                         // layer params (excluding "group" and "visibility")
      }
    ]
  },
  "defaultScales": [50000, 25000, 10000, 5000],     // required, default map scales
  "defaultPrintScales": [50000, 25000, 10000, 5000],// optional, confined list of available print scales. If not specified, scale is freely choosable.
  "defaultPrintResolutions": [150, 300, 600],       // optional, confined list of abailable print resolutions. If not specified, resolution is freely choosable.
  "defaultPrintGrid": [<as printGrid above>]        // optional, list of grid intervals to use for various scales when printing, no grid is primted if omitted
}
```

`yarn run themesconfig` requests the capabilities of all WMS services and
generates a runtime configuration `themes.json`. It is automatically run on
`yarn start`. To update the `themes.json` of a deployed application, just run
`yarn run themesconfig` in the development tree and copy the resulting
`themes.json` to the production folder.

### URL parameters
The following parameters can appear in the URL:

 * `t`: The active theme
 * `l`: The visible layers in the map
 * `bl`: The visible background layer
 * `st`: The search text
 * `e`: The visible extent
 * `c`: The center of the visible extent
 * `s`: The current scale
 * `crs`: The CRS of extent/center coordinates

The `urlPositionFormat` parameter in `config.json` determines whether extent or
center + scale appear in the URL. Allowed values are `extent` and `centerAndZoom`.
Default is `extent`.

The `urlPositionCrs` parameter in `config.json` determines the CRS to use for the
extent/center coordinates in the URL. The value is an EPSG code (i.e. `EPSG:4326`).
Default is the map projection.

If `urlPositionCrs` is equal to the map projection, the `crs` parameter is
omitted in the URL, but it can be manually added to the URL to specify the initial
extent/center in a desired CRS.

### Server

A server component is necessary for generating and resolving permalinks.
A sample server can be found [here](https://github.com/sourcepole/qwc2-server).
The url to the server must be specified under `qwc2serverUrl` in `config.json`.

Translations
------------

Translations are stored in the `data.<locale>` files stored under `translations`.
These are plain text files in JSON format. Translations available to the application
need to be listed under `supportedLocales` in `js/appConfig.js`.

When running `yarn start` (or manually `yarn run tsupdate`), translations from
the common QWC2 components, located in `qwc2/translations`, are automatically
merged. This ensures the translations remain up-to-date when updating the `qwc2`
submodule. The languages for which translations need to be generated are listed
under `translations/tsconfig.json`. If you ship applications-specific components,
you must list the message ids in the `strings` section of `translations/tsconfig.json`,
analogously to `qwc2/translations/tsconfig.json`.

You can test a locale by adding the `locale` query parameter to the URL, i.e. `locale=de`.

Deployment
----------

Run

    yarn run prod

The files created inside `./prod` can then be deployed to a web server.

If you deploy to a folder other than the root of the web server, be sure to adjust
the asset paths in `styleConfig.js` and `config.json`.


HTTP headers for printing
-------------------------

QGIS Server doesn't emit a HTTP header with a file name, which usually results in an ugly file name for print outputs. Therefore the web server should add a `Content-Disposition` header.

Example for Apache:

    SetEnvIf Request_URI "^/wms.*/(.+)$" project_name=$1
    Header always setifempty Content-Disposition "attachment; filename=%{project_name}e.pdf" "expr=%{CONTENT_TYPE} = 'application/pdf'"

This example uses the `setenvif` and the `headers` modules:

    a2enmod setenvif
    a2enmod headers


Preparation for developers
--------------------------

### Short MapStore2 introduction

* https://github.com/geosolutions-it/MapStore2/wiki/ReactJS-and-Redux-introduction

### ECMAScript 2015

* https://babeljs.io/docs/learn-es2015/

### ReactJS

* https://facebook.github.io/react/docs/thinking-in-react.html
* https://medium.com/@diamondgfx/learning-react-with-create-react-app-part-1-a12e1833fdc
* https://medium.com/@diamondgfx/learning-react-with-create-react-app-part-2-3ad99f38b48d
* https://medium.com/@diamondgfx/learning-react-with-create-react-app-part-3-322447d14192


### Redux

* http://redux.js.org/
* https://egghead.io/courses/getting-started-with-redux
* https://egghead.io/courses/building-react-applications-with-idiomatic-redux


ChangeLog.md
v0.3 (XXX XX 2017):

**Important**:
- Run `yarn install` to upgrade dependencies.
- Consult the list of incompatible changes below.

**News**:
- Allow specifying WMS image format in themes configuration
- Add Toolbar component for quick-launch buttons next to search bar
- Add option to store current view as center + scale in URL
- Add support for URL crs parameter to specify in wich CRS the center/extent coordinates are specified
- Make TopBar components configurable
- Make logo format configurable
- Move coordinates search provider to app
- Add optional search provider selection menu
- Add support for search queries with multiple input fields
- Allow removing entire layers from identify results list
- Add button to export the identify results to json
- Add component to identify all features within a selected region
- Add possibility to not display any background layer
- Make the short permalink share component fall back to the full URL if no QWC2 server is configured
- Make map button positions configurable
- Add button to toggle the layertree
- Add expanders to layer tree groups
- Allow hiding layer icons in layertree
- Honour layer drawing order
- Allow limiting available map scales for printing
- Also encode layer transparencies in the layer query parameter
- Make the overview map display the actual theme map instead of an OSM layer
- Add info dialog for displaying layer metadata and legend
- Expose composer labels in print dialog
- Show label with search result text along with pin marker
- Also highlight the geometry of the selected search result in the print output
- Increase size of theme thumbnails
- Add raster-export component to save a selected area of the map to an image
- Add button to print current layertree legend
- Automatically select search result if only one result is returned
- Automatically perform search when theme is loaded an a search text query parameter exists. If the search results in exactly one result, it will get selected.
- Make custom commands in package.json windows compatible
- Update all dependencies to latest version available as of Feb 27 2017
- Add draw component
- Move various scripts to qwc2/scripts
- Allow selectively overriding translation strings while importing the rest from the qwc2/translations
- Map projections are now set individually for each theme
- Add preserveExtentOnThemeSwitch and preserveBackgroundOnThemeSwitch to config.json to control whether extent / background layer should be preserved (if possible) on theme switch
- Make identify results dialog resizeable
- Insert anchor tags for urls / email addresses in feature attributes
- Allow configuring grid interval for print output
- Allow specifying human readable labels for the CRS EPSG codes in js/appConfig.js
- Allow limiting available resolutions for printing
- Add per-theme "collapseLayerGroupsBelowLevel" setting to control level below which layer groups are collapsed by default in the layer tree
- Add per-theme "skipEmptyFeatureAttributes" setting to control whether empty/NULL attributes should be skipped in the feature info table
- Add optional label attribute to search result items to display in map on select (defaults to item text)
- Added LayerTree config properties: showRootEntry, showQueryableIcon, allowMapTips
- Use assets/templates/legendprint.html as a template for the map legend print.
- Added LayerTree groupTogglesSublayers config property to control whether toggling a group toggles just the group or also all the sublayers

**Incompatible changes**:
- Keys of TopBar menuItems in config.js have been renamed
- Center/extent coordinates in URL by default in map projection instead of EPSG:4326
- The search text URL parameter key is now `st` instead of `s`
- The individual TopBar plugin components are now configured in js/appConfig.js
- The coordinate search provider was moved to js/SearchProviders.js
- The search providers onSearch and getMoreResults take an additional requestId argument
- The searchResultsLoaded action has changed:

        -const {searchResultLoaded} = require("../qwc2/MapStore2/web/client/actions/search");
        +const {addSearchResults} = require("../qwc2/QWC2Components/actions/search");

- The search providers must return the additional fields provider and reqId:

        {data: results, provider: providerId, reqId: requestId}

- Various scripts have been moved to qwc2/scripts. The "scripts" section of package.json has been adapted to use these.
- The translations/updateTranslations.py was removed in favour of qwc2/scripts/updateTranslations.js, the translation  languages now need to be specified in translations/tsconfig.json along with application specific message ids. Consult the README for more details.
- Map projections are now set individually for each theme via the mapCrs entry. If unset, it will default to EPSG:3857.
- The defaultScales list in themesConfig.json is now required
- The locale definitions under `supportedLocales` in `appConfig.js` must provide a localeData field

        localeData: require('react-intl/locale-data/<lang>')

- The print legend functionality requires a assets/templates/legendprint.html file with an element with id="legendcontainer", which will contain the legend graphics. The HTML document title of this template **may** influence the suggested output name when printing to a file (depending on the browser).


v0.2 (Jan 03 2017):
- https://github.com/qgis/qwc2-demo-app/compare/v0.1...v0.2


v0.1 (Dec 21 2017):
- Initial public release


themesConfig.js
/**
 * Copyright 2016, Sourcepole AG.
 * All rights reserved.
 *
 * This source code is licensed under the BSD-style license found in the
 * LICENSE file in the root directory of this source tree.
 */

/*eslint no-console: 0, vars-on-top: 0, camelcase: 0*/

const urlUtil = require('url');
const axios = require('axios');
const xml2js = require('xml2js');
const fs = require('fs');

// load thumbnail from file or GetMap
function getThumbnail(configItem, resultItem, layers, crs, extent, resolve) {
    if (configItem.thumbnail !== undefined) {
        // check if thumbnail can be read
        if(fs.existsSync("./assets/img/mapthumbs/" + configItem.thumbnail)) {
            resultItem.thumbnail = "img/mapthumbs/" + configItem.thumbnail;
            // finish task
            resolve(true);
            return;
        }
    }

    console.error("Using WMS GetMap to generate thumbnail for " + configItem.url);

    // WMS GetMap request
    var parsedUrl = urlUtil.parse(configItem.url, true);
    parsedUrl.search = '';
    parsedUrl.query.SERVICE = "WMS";
    parsedUrl.query.VERSION = "1.3.0";
    parsedUrl.query.REQUEST = "GetMap";
    parsedUrl.query.FORMAT = "image/png";
    parsedUrl.query.TRANSPARENT = "TRUE";
    parsedUrl.query.STYLES = "";
    parsedUrl.query.WIDTH = 128;
    parsedUrl.query.HEIGHT = 96;
    parsedUrl.query.CRS = crs;
    let bboxw = extent[2] - extent[0];
    let bboxh = extent[3] - extent[1];
    let bboxcx = 0.5 * (extent[0] + extent[2]);
    let bboxcy = 0.5 * (extent[1] + extent[3]);
    let imgratio = 128./96.;
    if(bboxw > bboxh) {
        let bboxratio = bboxw/bboxh;
        if(bboxratio > imgratio) {
            bboxh = bboxw / imgratio;
        } else {
            bboxw = bboxh * imgratio;
        }
    } else {
        bboxw = bboxh * imgratio;
    }
    let adjustedExtent = [bboxcx - 0.5 * bboxw, bboxcy - 0.5 * bboxh,
                          bboxcx + 0.5 * bboxw, bboxcy + 0.5 * bboxh];
    parsedUrl.query.BBOX = adjustedExtent.join(',');
    parsedUrl.query.LAYERS = layers.join(',');
    const getMapUrl = urlUtil.format(parsedUrl);

    axios.get(getMapUrl, {responseType: "arraybuffer"}).then((response) => {
        let basename = configItem.url.replace(/.*\//, "") + ".png";
        fs.writeFileSync("./assets/img/mapthumbs/" + basename, response.data);
        resultItem.thumbnail = "img/mapthumbs/" + basename;
        // finish task
        resolve(true);
    }).catch((error) => {
        console.error("ERROR for WMS " + configItem.url + ":\n", error);
        resultItem.error = "Could not get thumbnail";
        // finish task
        resolve(false);
    });
}

// convert non-array object to array containing the object
// used to restore arrays lost by 'explicitArray: false' xml2js option
function toArray(obj) {
    if (!Array.isArray(obj)) {
        return [obj];
    }
    return obj;
}

// recursively get layer tree
function getLayerTree(layer, resultLayers, visibleLayers, printLayers) {
    if (printLayers.indexOf(layer.Name) !== -1) {
        // skip print layers
        return;
    }

    var layerEntry = {
        name: layer.Name,
        title: layer.Title
    };
    if (layer.Layer === undefined) {
        if (layer.$.geometryType == "WKBNoGeometry") {
            // skip layers without geometry
            return;
        }

        // layer
        layerEntry.visibility = layer.$.visible === '1';
        if (layerEntry.visibility) {
            // collect visible layers
            visibleLayers.push(layer.Name);
        }
        layerEntry.queryable = layer.$.queryable === '1';
        if (layerEntry.queryable) {
            layerEntry.displayField = layer.$.displayField;
        }
        if (layer.Attribution !== undefined) {
            layerEntry.attribution = layer.Attribution.Title;
            if (layer.Attribution.OnlineResource !== undefined) {
                layerEntry.attributionUrl = layer.Attribution.OnlineResource.$['xlink:href'];
            }
        }
        layerEntry.opacity = 255;
        if (layer.MinScaleDenominator !== undefined) {
            layerEntry.minScale = parseInt(layer.MinScaleDenominator, 10);
            layerEntry.maxScale = parseInt(layer.MaxScaleDenominator, 10);
        }
        // use geographic bounding box, as default CRS may have inverted axis order with WMS 1.3.0
        layerEntry.crs = "EPSG:4326";
        layerEntry.extent = [
            parseFloat(layer.EX_GeographicBoundingBox.westBoundLongitude),
            parseFloat(layer.EX_GeographicBoundingBox.southBoundLatitude),
            parseFloat(layer.EX_GeographicBoundingBox.eastBoundLongitude),
            parseFloat(layer.EX_GeographicBoundingBox.northBoundLatitude)
        ];
    } else {
        // group
        layerEntry.sublayers = [];
        for (var subLayer of toArray(layer.Layer)) {
            getLayerTree(subLayer, layerEntry.sublayers, visibleLayers, printLayers);
        }
        if (layerEntry.sublayers.length === 0) {
            // skip empty groups
            return;
        }
    }
    resultLayers.push(layerEntry);
}

// parse GetCapabilities for theme
function getTheme(configItem, resultItem) {
    resultItem.url = configItem.url;

    var parsedUrl = urlUtil.parse(configItem.url, true);
    parsedUrl.search = '';
    parsedUrl.query.SERVICE = "WMS";
    parsedUrl.query.VERSION = "1.3.0";
    parsedUrl.query.REQUEST = "GetProjectSettings";
    const getCapabilitiesUrl = urlUtil.format(parsedUrl);

    return new Promise((resolve) => {
        axios.get(getCapabilitiesUrl).then((response) => {
            // parse capabilities
            var capabilities;
            xml2js.parseString(response.data, {
                tagNameProcessors: [xml2js.processors.stripPrefix],
                explicitArray: false
            },
            (ignore, result) => {
                if (result === undefined || result.WMS_Capabilities === undefined) {
                    // show response data on parse error
                    throw new Error(response.data);
                } else {
                    capabilities = result.WMS_Capabilities;
                }
            });

            console.log("Parsing WMS GetProjectSettings of " + configItem.url);

            const topLayer = capabilities.Capability.Layer;

            const themeId = topLayer.Name;

            // use name from config or fallback to WMS title
            const wmsTitle = configItem.title || capabilities.Service.Title || topLayer.Title;

            // keywords
            var keywords = [];
            toArray(capabilities.Service.KeywordList.Keyword).map((entry) => {
                var value = (typeof entry === 'object') ? entry._ : entry;
                if (value !== "infoMapAccessService") {
                    keywords.push(value);
                }
            });

            // use first CRS for thumbnail request
            const crs = toArray(topLayer.CRS).filter(item => item != 'CRS:84')[0];
            var extent = [];
            for (var bbox of toArray(topLayer.BoundingBox)) {
                if (bbox.$.CRS === crs) {
                    extent = [
                        parseFloat(bbox.$.minx),
                        parseFloat(bbox.$.miny),
                        parseFloat(bbox.$.maxx),
                        parseFloat(bbox.$.maxy)
                    ];
                    break;
                }
            }

            // collect WMS layers for printing
            var printLayers = [];
            if (configItem.backgroundLayers !== undefined) {
                printLayers = configItem.backgroundLayers.map((entry) => {
                    return entry.printLayer;
                });
            }

            // layer tree and visible layers
            var layerTree = [];
            var visibleLayers = [];
            getLayerTree(topLayer, layerTree, visibleLayers, printLayers);
            visibleLayers.reverse();

            // print templates
            var printTemplates = [];
            if (capabilities.Capability.ComposerTemplates && capabilities.Capability.ComposerTemplates.length > 0) {
                for (var composerTemplate of capabilities.Capability.ComposerTemplates.ComposerTemplate) {
                    var printTemplate = {
                        name: composerTemplate.$.name
                    };
                    if (composerTemplate.ComposerMap !== undefined) {
                        // use first map from GetProjectSettings
                        var composerMap = toArray(composerTemplate.ComposerMap)[0];
                        printTemplate.map = {
                            name: composerMap.$.name,
                            width: parseFloat(composerMap.$.width),
                            height: parseFloat(composerMap.$.height)
                        };
                    }
                    if (composerTemplate.ComposerLabel !== undefined) {
                        printTemplate.labels = toArray(composerTemplate.ComposerLabel).map((entry) => {
                            return entry.$.name;
                        });
                    }
                    printTemplates.push(printTemplate);
                }
            }

            // update theme config
            resultItem.id = themeId;
            resultItem.name = topLayer.Name;
            resultItem.title = wmsTitle;
            resultItem.attribution = configItem.attribution;
            resultItem.attributionUrl = configItem.attributionUrl;
            resultItem.keywords = keywords.join(', ');
            resultItem.format = configItem.format;
            resultItem.tiled = configItem.tiled;
            // use geographic bounding box for theme, as default CRS may have inverted axis order with WMS 1.3.0
            resultItem.crs = "EPSG:4326";
            resultItem.extent = [
                parseFloat(topLayer.EX_GeographicBoundingBox.westBoundLongitude),
                parseFloat(topLayer.EX_GeographicBoundingBox.southBoundLatitude),
                parseFloat(topLayer.EX_GeographicBoundingBox.eastBoundLongitude),
                parseFloat(topLayer.EX_GeographicBoundingBox.northBoundLatitude)
            ];
            resultItem.scales = configItem.scales;
            // NOTE: skip root WMS layer
            resultItem.sublayers = layerTree[0].sublayers;
            resultItem.backgroundLayers = configItem.backgroundLayers;
            resultItem.searchProviders = configItem.searchProviders;
            resultItem.additionalMouseCrs = configItem.additionalMouseCrs;
            if (printTemplates.length > 0) {
                resultItem.print = printTemplates;
            }

            // set default theme
            if (configItem.default) {
                result.themes.defaultTheme = resultItem.name;
            }

            // get thumbnail asynchronously
            getThumbnail(configItem, resultItem, visibleLayers, crs, extent, resolve);
        }).catch((error) => {
            console.error("ERROR reading WMS GetProjectSettings of " + configItem.url + ":\n", error);
            resultItem.error = "Could not read GetProjectSettings";
            resultItem.title = "Error";
            // finish task
            resolve(false);
        });
    });
}

// asynchronous tasks
var tasks = [];

// recursively get themes for groups
function getGroupThemes(configGroup, resultGroup) {
    for (var item of configGroup.items) {
        var itemEntry = {};
        tasks.push(getTheme(item, itemEntry));
        resultGroup.items.push(itemEntry);
    }

    if (configGroup.groups !== undefined) {
        for (var group of configGroup.groups) {
            var groupEntry = {
                title: group.title,
                items: [],
                subdirs: []
            };
            getGroupThemes(group, groupEntry);
            resultGroup.subdirs.push(groupEntry);
        }
    }
}

/* load themesConfig.json:
  {
    "themes": {
      "items": [
        {
          "url": "<http://localhost/wms/theme>",
          "title": "<Custom theme title>",            // optional, use WMS title if not set
          "thumbnail": "<filename>",                  // optional image file in assets/img/mapthumbs, if not set uses WMS GetMap to generate the thumbnail and stores it in assets/img/mapthumbs
          "attribution": "<Attribution>",             // optional theme attribution
          "attributionUrl": "<attribution URL>",      // optional theme attribution URL
          "default": true,                            // optional, set this as the initial theme
          "scales": [25000, 10000, 5000, 2500],       // optional custom map scales
          "tiled": true,                              // optional, use tiled WMS (default is false)
          "format": "image/png",                      // optional, the image format to use in the WMS request, defaults to image/png
          "backgroundLayers": [                       // optional background layers
            {
              "name": "<background layer name>",      // background layer name from list below
              "printLayer": "<WMS layer name>",       // optional equivalent WMS layer name for printing
              "visibility": true                      // optional initial visibility on topic selection
            }
          ],
          "searchProviders": ["<search provider>"],   // optional search providers
          "additionalMouseCrs": ["<epsg code>"]       // optional list of additional CRS for mouse position (map projection and WGS84 are listed by default). Make sure proj defs are loaded in js/appConfig.js.
        }
      ],
      "groups": [                                     // optional, nested groups
        {
          "title": "<Group title>",
          "items": [
            {
              "url": "<http://localhost/wms/group_theme>"
            }
          ]
          "groups": [
            // subgroups
          ]
        }
      ],
      "backgroundLayers": [                           // optional list of background layers for themes
        {
          "name": "<background layer name>",          // referenced by themes
          "title": "<Background layer title>",
          "thumbnail": "<filename>",                  // optional image file in assets/img/mapthumbs, use default.jpg if not set
          ...                                         // layer params (excluding "group" and "visibility")
        }
      ]
    },
    "defaultScales": [50000, 25000, 10000, 5000]      // optional default map scales
  }
*/
console.log("Reading themesConfig.json");
var config = require('./themesConfig.json');

var result = {
    themes: {
        title: "root",
        subdirs: [],
        items: [],
        defaultTheme: undefined,
        defaultScales: config.defaultScales,
        backgroundLayers: config.themes.backgroundLayers
    }
};
getGroupThemes(config.themes, result.themes);

if (result.themes.backgroundLayers !== undefined) {
    // get thumbnails for background layers
    result.themes.backgroundLayers.map((backgroundLayer) => {
        let imgPath = "img/mapthumbs/" + backgroundLayer.thumbnail;
        if (!fs.existsSync("./assets/" + imgPath)) {
            imgPath = "img/mapthumbs/default.jpg";
        }
        backgroundLayer.thumbnail = imgPath;
    });
}

Promise.all(tasks).then(() => {
    // write config file
    fs.writeFile('./themes.json', JSON.stringify(result, null, 2), (error) => {
        if (error) {
            console.error("ERROR:", error);
        } else {
            console.log("\nCreated themes.json\n\n");
        }
    });
}).catch((error) => {
    console.error("ERROR:", error);
});
